R/standardize_subject_ids.R

```diff
@@ -0,0 +1,213 @@
+#'
+#' @examples
+#' dat <- check_subject_ids(
+#'   data           = readRDS(system.file("extdata", "test_df.RDS",
+#'                                        package = "cleanepi")),
+#'   target_columns = "study_id",
+#'   prefix         = "PS",
+#'   suffix         = "P2",
+#'   range          = c(1, 100),
+#'   nchar          = 7
+#' )
+#' @export
+check_subject_ids <- function(data,
+                              target_columns,
+                              prefix         = NULL,
+                              suffix         = NULL,
+                              range          = NULL,
+                              nchar          = NULL) {
+  checkmate::assert_data_frame(data, null.ok = FALSE)
+  checkmate::assert_character(target_columns, null.ok = FALSE,
+                              any.missing = FALSE, len = 1L)
+  checkmate::assert_character(prefix, len = 1L, null.ok = TRUE,
+                              any.missing = FALSE)
+  checkmate::assert_character(suffix, len = 1L, null.ok = TRUE,
+                              any.missing = FALSE)
+  checkmate::assert_vector(range, any.missing = FALSE, min.len = 2L,
+                           null.ok = TRUE, unique = TRUE, max.len = 2L)
+  checkmate::assert_numeric(nchar, null.ok = TRUE, any.missing = FALSE,
+                            len = 1L)
+
+  # coerce id column to character
+  if (is.numeric(data[[target_columns]]) | is.factor(data[[target_columns]])) {
+    data[[target_columns]] <- as.character(data[[target_columns]])
+  }
+
+  # check for missing and duplicated ids
+  data     <- check_subject_ids_oness(data, target_columns)
+  bad_rows <- NULL
+
+  # check prefix of subject IDs
+  if (!is.null(prefix)) {
+    prefix_check    <- startsWith(data[[target_columns]], prefix)
+    idx             <- which(!(prefix_check))
+    if (length(idx) > 0L) {
+      bad_rows      <- c(bad_rows, idx)
+    }
+  }
+
+  # check suffix of subject IDs
+  if (!is.null(suffix)) {
+    suffix_check <- endsWith(data[[target_columns]], suffix)
+    idx          <- which(!(suffix_check))
+    if (length(idx) > 0L) {
+      bad_rows   <- c(bad_rows, idx)
+    }
+  }
+
+  # detect subject IDs that do not match the provided format
+  if (!is.null(nchar)) {
+    length_check <- as.logical(as.character(lapply(data[[target_columns]],
+                                                   check_subject_ids_length,
+                                                   nchar)))
+    idx          <- which(length_check)
+    if (length(idx) > 0L) {
+      bad_rows   <- c(bad_rows, idx)
+    }
+  }
+
+  # check the numbers in the sample IDs
+  if (!is.null(range)) {
+    numbers_in <- as.numeric(unlist(lapply(data[[target_columns]],
+                                           readr::parse_number)))
+    idx        <- which(!(numbers_in >= min(range) & numbers_in <= max(range)))
+    if (length(idx) > 0L) {
+      bad_rows <- c(bad_rows, idx)
+    }
+  }
```

Comment:
In terms of performance, it would probably be better to construct a regex from the specified (or unspecified) arguments and apply it directly to the input

Reply:
I have made used of regex for the `prefix` and `suffix`.

It is difficult to anticipate on the `range` and elaborate an automated regex for it.

These changes can be found in commit `f7595f8`.